File: Sandbox/src/ExampleGame/Level.cpp
#include "Level.h"

using namespace Niking2D;
void Level::Init()
{
	m_TriangleTexture = Texture2D::Create("assets/textures/Triangle.png");

	m_Player.LoadAssets();

	m_Pillars.resize(5);
	for (int i = 0; i < 5; i++)
		CreatePillar(i, i * 10.0f);
}

void Level::OnUpdate(Niking2D::Timestep ts)
{
	m_Player.OnUpdate(ts);

	if (m_Player.GetPosition().x > m_PillarTarget)
	{
		CreatePillar(m_PillarIndex, m_PillarTarget + 20.0f);
		m_PillarIndex = ++m_PillarIndex % m_Pillars.size();
		m_PillarTarget += 10.0f;
	}

}

void Level::OnRender()
{
	const auto& playerPos = m_Player.GetPosition();

	glm::vec4 color = glm::vec4(1.0f);
	//	Background
	Renderer2D::DrawQuad({ playerPos.x, 0.0f, -0.8f }, { 50.0f, 50.0f }, { 0.3f, 0.3f, 0.3f, 1.0f });

	// Floor and ceiling
	Renderer2D::DrawQuad({ playerPos.x,  34.0f }, { 50.0f, 50.0f }, color);
	Renderer2D::DrawQuad({ playerPos.x, -34.0f }, { 50.0f, 50.0f }, color);

	for (auto& pillar : m_Pillars)
	{
		Renderer2D::DrawQuad(pillar.TopPosition, pillar.TopScale, glm::radians(180.0f), m_TriangleTexture, 1.0f, color);
		Renderer2D::DrawQuad(pillar.BottomPosition, pillar.BottomScale, 0.0f, m_TriangleTexture, 1.0f, color);
	}



	m_Player.OnRender();
}

void Level::OnImGuiRender()
{
	m_Player.OnImGuiRender();
}

void Level::CreatePillar(int index, float offset)
{
	Pillar& pillar = m_Pillars[index];
	pillar.TopPosition.x = offset;
	pillar.BottomPosition.x = offset;
	pillar.TopPosition.z = index * 0.1f - 0.5f;
	pillar.BottomPosition.z = index * 0.1f - 0.5f + 0.05f;

	float center = Random::Float() * 35.0f - 17.5f;
	float gap = 2.0f + Random::Float() * 5.0f;

	pillar.TopPosition.y = 10.0f - ((10.0f - center) * 0.2f) + gap * 0.5f;
	pillar.BottomPosition.y = -10.0f - ((-10.0f - center) * 0.2f) - gap * 0.5f;


}

File: Sandbox/src/ExampleScript/level.lua
level = {}
level.pillarTarget = 30.0
level.pillarIndex = 0

pillar = {}
pillar.sprite = sprites.triangle
pillar.topposition = {x = 0.0, y= 10.0, z= 0.0}
pillar.topscale = {x = 15.0, y = 20.0}

pillar.bottomposition = {x = 10.0, y= 10.0, z= 0.0}
pillar.bottomscale = { x = 15.0, y = 20.0}



pillars = {}

function level:Init()
    for i=1,5 do
        pillars[i] = {topposition = {}, bottomposition = {},bottomscale = {x = 15.0, y = 20.0}, topscale = {x = 15.0, y = 20.0} }
        CreatePillar(pillars[i], i, (i-1) * 10.0);

        print(player.position.x, player.position.y,player:GetRotation(),
        pillars[i].topposition.x, pillars[i].topposition.y,
        pillars[i].topscale.x, pillars[i].topscale.y)
    end
end

function level:Update()
    if player.position.x > level.pillarTarget then
        CreatePillar(pillars[level.pillarIndex + 1],level.pillarIndex, level.pillarTarget + 20.0)
        level.pillarIndex = (level.pillarIndex + 1) % 5
        level.pillarTarget = level.pillarTarget + 10.0 
    end

    for i=1,5 do
        if IsCollision(player.position.x, player.position.y,player:GetRotation(),
        pillars[i].topposition.x, pillars[i].topposition.y,
        pillars[i].topscale.x, pillars[i].topscale.y, 180.0) == true 
        or 
            IsCollision(player.position.x, player.position.y,player:GetRotation(),
            pillars[i].bottomposition.x, pillars[i].bottomposition.y,
            pillars[i].bottomscale.x, pillars[i].bottomscale.y, 0.0) == true
        then
            level:GameOver()
            return
        end
    end
end

function level:GameOver()
    GameState.State = GameState.GameOver
end

function level:Draw()
    local color = {r = 1.0, g = 1.0, b = 1.0, a = 1.0}
	-- Background
	DrawQuadZ(player.position.x, 0.0, -0.8, 50.0, 50.0, 0.3,0.3,0.3, 1.0)
	-- Floor and ceiling
	DrawQuad(player.position.x, 34.0, 50.0, 50.0, color.r, color.g, color.b, color.a)
    DrawQuad(player.position.x, -34.0, 50.0, 50.0, color.r, color.g, color.b, color.a)
    
    for i=1,5 do
        -- DrawQuadRTex(sprites.player, player.position.x, player.position.y, 1.0, 1.3, math.rad(GetRotation()))
        DrawQuadZRTex(pillar.sprite, pillars[i].topposition.x, pillars[i].topposition.y, pillars[i].topposition.z, pillars[i].topscale.x,  pillars[i].topscale.y, math.rad(180.0))

        DrawQuadZRTex(pillar.sprite, pillars[i].bottomposition.x,pillars[i].bottomposition.y, pillars[i].bottomposition.z, pillars[i].bottomscale.x,  pillars[i].bottomscale.y, 0)
    end
end

function CreatePillar(p, index, offset)
    local pillar = p
    pillar.topposition.x = offset
	pillar.bottomposition.x = offset
	pillar.topposition.z = (index-1) * 0.1 - 0.5
	pillar.bottomposition.z = (index-1) * 0.1 - 0.5 + 0.05

	local center = math.random(100) / 100 * 35.0 - 17.5
    local gap = 2.0 + math.random(100) / 100 * 5.0

	pillar.topposition.y = 10.0 - ((10.0 - center) * 0.2) + gap * 0.5
	pillar.bottomposition.y = -10.0 - ((-10.0 - center) * 0.2) - gap * 0.5
end